Adding basic support for int mod and div operations

blishko opened the pull request:
I am opening this PR draft to discuss changes necessary to add support for `div` and `mod` in LIA logic.

So far, there is support for creating `div` and `mod` terms where the second argument is constant. 
A utility to rewrite formulas with these operations to equivalent formulas without them is still needed.

blishko commented on src/logics/LIALogic.h:
```diff
@@ -32,6 +34,8 @@ class LIALogic: public LALogic
     static const char*  tk_int_times;
     static const char*  tk_int_div;
+    static const char*  tk_int_mod;
+    static const char*  tk_int_abs;
```
I noticed that `Int` theory declares also `abs` function, so I put it here, but maybe I can remove it since it is not necessary for what I need right now.

blishko commented:
@aehyvari sorry for all the changes, but there was some inconsistencies in how `LALogic` delegates the queries to `LIALogic` and `LRALogic`, so I did some refactoring.

blishko commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+
+std::size_t DivModRewriter::counter = 0;
+const char * DivModRewriter::divPrefix = ".div";
+const char * DivModRewriter::modPrefix = ".mod";
```
I thought about if this should be in LIALogic or here in the rewriter. Now I am more inclined  to keep it here.
It feels closer to there it is actually used and I believe it should not cause any problems.

blishko commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+    auto it = substitutions.find(root);
+    assert(it != substitutions.end());
+    definitions.push(it->second);
+    return logic.mkAnd(definitions);
```
While writing this function, I realized it is very close to `Logic::simplifyTree`.
I think it would be good to unify the approach.
One idea would be to implement this as part of `LIA::visit`, but I have a different idea that I like better.
We would have a general rewriter (living outside of `Logic`) which would be rewriting the DAG based on registered plugins. The solver than can customize the rewriter for different theories, but it would also allow users to create their own rewriters.

aehyvari commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+                toProcess.push_back(DFSEntry(nextChild));
+            }
+            continue;
+        }
```
I've been writing the above code slightly different, see for instance [the corresponding loop in IteToSwitch.cc](https://github.com/usi-verification-and-security/opensmt/blob/e33b1470696e02e9360f4091ceadfdf8b709657d/src/itehandler/IteToSwitch.cc#L40).  Not sure which one is better, but I guess we should settle for one version.  I think grepping `unprocessed_children` would reveal most of these cases.

aehyvari replied:
This is probably an argument for introducing the general visit pattern.  Or maybe that's yet another topic since we don't do the rewriting in all of the cases.

blishko replied:
I am thinking that we could have a const visitor that does not modify the terms, just collects information and than separately a rewriter that rewrites the terms.

aehyvari commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+        auto & currentEntry = toProcess.back();
+        PTRef currentRef = currentEntry.term;
+        Pterm const & term = logic.getPterm(currentRef);
+        unsigned childrenCount = term.size();
```
size_t would maybe work better below.

aehyvari commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+        assert(done[Idx(term.getId())] == 0);
+        vec<PTRef> newArgs(childrenCount);
+        bool needsChange = false;
+        for (unsigned i = 0; i < childrenCount; ++i) {
```
`unsigned` => `auto` (at least if `childrenCount` is `size_t` this works)

aehyvari commented on src/simplifiers/LARewriter.h:
```diff
@@ -0,0 +1,33 @@
+        PTRef div;
+        PTRef mod;
+    };
+    std::unordered_map<std::pair<PTRef, PTRef>, DivModPair, PTRefPairHash> divModeCache;
```
Should this be called `divModCache` instead of `divModeCache`?

blishko replied:
Fixed.

aehyvari commented on src/logics/LIALogic.h:
```diff
@@ -47,59 +51,51 @@ class LIALogic: public LALogic
+
+    const PTRef mkIntMod(PTRef first, PTRef second);
+    const PTRef mkIntAbs(PTRef arg);
+    PTRef mkNumDiv(vec<PTRef> const & args) override;
```
Why this takes a `vec` and `mkIntMod` takes two `PTRef`s?

blishko replied:
I think this is how it was before. When it was common for LIA and LRA.

Should we distinguish between the real operator `/` and the integer operator `div` at the level of `mk` methods?

aehyvari replied:
This was already resolved.

aehyvari commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+    return DivModRewriter().rewriteDivMod(logic, root);
+}
+
+// TODO: figure out how to cache assigned variables to avoid duplication
```
Is this comment still valid?  It looks like there is a cache already in place, but maybe it refers to something else?

blishko replied:
I think it is dealt with now.

aehyvari reviewed (changes requested):
Looks excellent!  Is this code used somewhere outside of regression tests?

blishko commented on src/simplifiers/LARewriter.cc:
```diff
@@ -0,0 +1,111 @@
+    struct DFSEntry {
+        DFSEntry(PTRef term) : term(term) {}
+        PTRef term;
+        unsigned int nextChild = 0;
```
I thought that range of unsigned int is sufficient and it requires less memory than `size_t`.

blishko commented:
The code is not used anywhere outside the unit tests. I was thinking where to call it and came across the similarity with `Logic::visit`, and thought that I will try to unify them in a single mechanism.

Also, I think I have not yet added support in the parser for this.

blishko commented:
@aehyvari , I have rebased this on top of current master and rewritten it using the `Rewriter` template.
I am thinking of **not** using it in the preprocessing since it is not needed for the SMT-LIB benchmarks we actually support.
Rather, we can just keep it as a functionality available through API (that is enough for my use case).

I am also thinking of hiding `DivModConfig` inside `DivModRewriter`, but I can do it later, when addressing your anticipated comments.

aehyvari commented on src/rewriters/DivModRewriter.h:
```diff
@@ -0,0 +1,111 @@
+    vec<PTRef> definitions;
+
+    const char * divPrefix = ".div";
+    const char * modPrefix = ".mod";
```
Define the strings in LIATheory?

blishko replied:
I think this would create a dependency of `DivModRewriter` on `LIATheory` that does not really make sense. I would keep the rewriter as much independent as possible.
If these were to move somewhere, I would move them to `LIALogic`, that dependency already exists and is necessary.

aehyvari replied:
Ok, maybe we could move them there.

blishko replied:
I am not sure anymore about moving them out of here.
Do we plan to use them anywhere else?
I now see it more like an internal thing of the rewriter.

aehyvari replied:
Ok, as discussed, let's go with the current solution.

aehyvari commented on src/rewriters/DivModRewriter.h:
```diff
@@ -0,0 +1,111 @@
+                                logic.mkNumLeq(logic.getTerm_NumZero(), modVar),
+                                logic.mkNumLeq(modVar, logic.mkConst(upperBound))
+                        )
+                ));
```
What if we move `definitions` to LIATheory and have this stuff in a later phase.  Also the cache could live there, this could help to handle incrementality.

blishko replied:
I wanted to have independent rewriter that users (me) can use on their own.
I think we can achieve the same effect as what you suggest by having a single rewriter instance in the `LIATheory` and not recreate it for every call to the simplification.

aehyvari replied:
Ok, that sounds good.

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -249,3 +204,67 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+
+PTRef LIALogic::mkIntDiv(vec<PTRef> const & args) {
+    if (args.size() != 2) { throw OsmtApiException("Incorrect number of arguments for DIV operator"); }
+    return _mkIntDiv(args);
```
Div is defined as `(div Int Int Int :left-assoc)`.  Maybe it wouldn't be too much work to split it here into binary calls?

aehyvari replied:
We'll do this later in a separate PR.

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -249,3 +204,67 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+    PTRef divisor = args[1];
+    if (not isConstant(divisor)) {
+        throw LANonLinearException("Only division by constant is permitted in linear arithmetic!");
+    }
```
We should do the corresponding throws in `isNumZero(divisor)`

aehyvari commented on src/logics/LIALogic.cc:
```diff
@@ -249,3 +204,67 @@ PTRef LIALogic::sumToNormalizedInequality(PTRef sum) {
+    assert(args.size() == 2);
+    PTRef dividend = args[0];
+    PTRef divisor = args[1];
+    if (not isNumConst(divisor)) { throw OsmtApiException("Divisor must be constant in linear logic"); }
```
Throw when `isNumZero(divisor)`

blishko replied:
Added.

aehyvari commented on src/logics/LRALogic.cc:
```diff
@@ -131,22 +133,32 @@ LRALogic::LRALogic() :
+    vec<PTRef> args_new;
+    SymRef s_new;
+    if (args.size() != 2) {
+        throw OsmtApiException("Division operation requries exactly 2 arguments");
```
We could fix this as well to comply to the standard (` (/ Real Real Real :left-assoc)`)

blishko replied:
I am not sure I understand what that means. If I write `x / 2 / 3`, it should be read as `(x/2)/3`, right? This means we can accept more than two arguments, but each except the first one must be constant, right? Can we do it in another PR?

aehyvari replied:
Ok, another PR.

aehyvari commented on test/unit/test_ArithmeticExpressions.cpp:
```diff
@@ -46,7 +46,7 @@ TEST_F(ArithmeticExpressions_test, test_One_Arg_Multiplication)
     PTRef one = logic.mkConst("1");
     PTRef a = logic.mkNumVar("a");
-    PTRef div = logic.mkNumDiv(a, two);
+    PTRef div = logic.mkRealDiv(a, two);
```
We could maybe have unit tests for `LALogic::mkNumDiv` for both cases, where the instance is actually `LRALogic` and when it is `LIALogic`.

aehyvari reviewed (commented):
It looks like we're almost there, I just left a few comments!

blishko commented on src/logics/LALogic.h:
```diff
@@ -49,32 +49,32 @@ class LALogic: public Logic
 
 
     virtual const SymRef get_sym_Num_TIMES () const =0;
-    virtual const SymRef get_sym_Num_DIV () const =0;
```
Yes, I will bring this back.

blishko commented on src/logics/LALogic.h:
```diff
@@ -86,36 +86,37 @@ class LALogic: public Logic
-    bool isNumTimes(SymRef sr) const;
-    virtual bool isNumTimes(PTRef tr) const ;
-    bool isNumDiv(SymRef sr) const;
-    virtual bool isNumDiv(PTRef tr) const;
```
These are still available (just below). They are no longer virtual, but rather rely on the virtual `get_sym_Num_XXX` methods.

aehyvari replied:
This was resolved in gitter.

blishko commented on src/logics/LALogic.h:
```diff
@@ -150,7 +151,6 @@ class LALogic: public Logic
     virtual bool isLinearTerm(PTRef tr) const;
     virtual bool isLinearFactor(PTRef tr) const;
     virtual void splitTermToVarAndConst(const PTRef &term, PTRef &var, PTRef &fac) const;
-    virtual PTRef normalizeSum(PTRef sum);
```
Yes, `sumToNormalizedInequality` is doing that. Both get normalized to `0 <= z-x-y`. I'll check if we have tests for this.

blishko replied:
Added a test for that for both LIA and LRA

aehyvari commented:
Would it be a problem if another rewriter used the same symbols
unintentionally?

On Sat, 27 Mar 2021, 10:19 Martin Blicha, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/rewriters/DivModRewriter.h
> <https://github.com/usi-verification-and-security/opensmt/pull/221#discussion_r602694969>
> :
>
> > +
> +class LIALogic;
> +
> +class DivModConfig : public DefaultRewriterConfig {
> +    LIALogic & logic;
> +
> +    struct DivModPair {
> +        PTRef div;
> +        PTRef mod;
> +    };
> +
> +    std::unordered_map<std::pair<PTRef, PTRef>, DivModPair, PTRefPairHash> divModCache;
> +    vec<PTRef> definitions;
> +
> +    const char * divPrefix = ".div";
> +    const char * modPrefix = ".mod";
>
> I am not sure anymore about moving them out of here.
> Do we plan to use them anywhere else?
> I now see it more like an internal thing of the rewriter.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/usi-verification-and-security/opensmt/pull/221#discussion_r602694969>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB74VMUCODSGNG2JPXRC7CLTFWPLJANCNFSM4WZDLVAQ>
> .
>

blishko commented:
I have made a note about support for `/` and `div` with multiple arguments here: https://github.com/usi-verification-and-security/opensmt/issues/146#issuecomment-811860257

Seems we have a couple of operators that should support it.
